components/tools/OmeroPy/src/omero/gateway/__init__.py

```diff
@@ -1573,6 +1573,26 @@ def __init__(self, username=None, passwd=None, client_obj=None, group=None,
         # The properties we are setting through the interface
         self.setIdentity(username, passwd, not clone)
 
+    def __enter__(self):
+        """
+        Enter a context manager, connect if necesaary
+        Raise Exception if not connected
+
+        with BlitzGateway('user', 'password', host='host') as conn:
+            print list(conn.getObjects('Project'))
+        """
+        if not self._connected:
+            r = self.connect()
+            if not r:
+                raise Exception("Connect failed")
+        return self
+
+    def __exit__(self, *args):
+        """
+        Exit a context manager, close connection
+        """
+        self.close()
```

Comment:
Does close call the object even if not _connected?

Reply:
You can call close() as many times as you want without errors:
```
In [3]: conn._connected
Out[3]: True

In [4]: conn.close()

In [5]: conn._connected
Out[5]: False

In [6]: conn.close()

In [7]: conn._connected
Out[7]: False
```

Reply:
Sorry, that's not my worry, but rather if the `client_obj` is passed in the the _caller_ owns the object and not the `BlitzGateway` object.

Reply:
I can think of two reasonable options:
1. There's no point using `BlitzGateway(client_obj)` as a context manager unless you want it to be automatically closed, so we assume it's intentional, and that the user was passed the `client` from a source but is responsible for closing it.
2. Throw an exception if `BlitzGateway` is passed as a context manager and it's already connected.

Reply:
Option 1 is also consistent with the behaviour of the existing [`omero.cli.cli_login`](https://github.com/openmicroscopy/openmicroscopy/blob/v5.4.9/components/tools/OmeroPy/src/omero/cli.py#L1545).

1. `omero login ...`
2. Run a script that includes `with cli_login() as cli: ...`
3. When the script exists you will be logged out of the session created in step 1

Reply:
I can see getting behind Option 1. Can we just get the fact that the close will happen into the docs somewhere? I'd assume some example for exposing to users will need adding. cc: @pwalczysko @jburel

Reply:
@joshmoore so does that mean that with Option1 the lines like https://github.com/ome/training-scripts/blob/master/practical/groovy/analyse_particles_for_another_user.groovy#L218 will not be necessary anymore ? Maybe I am still missing the bigger picture here.

Reply:
What does that mean that a script ``exits`` ?

Reply:
> does that mean that with Option1 the lines like 

Yes, except at the moment, only in pure Python. We'd have to examine making BlitzGateway `Closeable` (if it's not already) on the Java side. cc: @dominikl 

> What does that mean that a script exits

In this case, it means when the `with` block ends:

```
with BlitzGateway() ...
   in the block
   in the block
   in the block
out of the block; no session
```

Reply:
Aha, so once this PR is merged, then the ``with BlitzGateway()...`` pattern is compulsory for all scripts ?  Not anymore https://github.com/ome/training-scripts/blob/master/practical/jython/analyse_particles_for_another_user.jy#L81 ?

Reply:
> pattern is compulsory for all scripts ?

Definitely not compulsory. This is just a convenience that was suggested by a workshop attendee in Kobe.

Reply:
Since we're agreed on the behaviour I've updated the docs: https://github.com/openmicroscopy/ome-documentation/pull/1915